cpp/oneapi/dal/algo/pca/backend/gpu/train_kernel_cov_impl_dpc.cpp

```diff
@@ -0,0 +1,730 @@
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#include "oneapi/dal/algo/pca/backend/gpu/train_kernel_cov_impl.hpp"
+#include "oneapi/dal/backend/common.hpp"
+#include "oneapi/dal/detail/common.hpp"
+#include "oneapi/dal/backend/primitives/ndarray.hpp"
+#include "oneapi/dal/detail/policy.hpp"
+#include "oneapi/dal/detail/profiler.hpp"
+#include "oneapi/dal/backend/memory.hpp"
+#include "oneapi/dal/backend/primitives/lapack.hpp"
+#include "oneapi/dal/backend/primitives/reduction.hpp"
+#include "oneapi/dal/backend/primitives/stat.hpp"
+#include "oneapi/dal/backend/primitives/blas.hpp"
+#include "oneapi/dal/backend/primitives/utils.hpp"
+#include "oneapi/dal/table/row_accessor.hpp"
+#include "oneapi/dal/algo/pca/backend/common.hpp"
+#include "oneapi/dal/algo/pca/backend/sign_flip.hpp"
+
+#ifdef ONEDAL_DATA_PARALLEL
+
+namespace oneapi::dal::pca::backend {
+
+namespace de = dal::detail;
+namespace bk = dal::backend;
+namespace pr = dal::backend::primitives;
+
+using alloc = sycl::usm::alloc;
+
+template <typename Data>
+using local_accessor_rw_t =
+    sycl::accessor<Data, 1, sycl::access::mode::read_write, sycl::access::target::local>;
+
+using comm_t = bk::communicator<spmd::device_memory_access::usm>;
+using dal::backend::context_gpu;
+using method_t = method::cov;
+using model_t = model<task::dim_reduction>;
+using task_t = task::dim_reduction;
+using input_t = train_input<task_t>;
+using result_t = train_result<task_t>;
+using descriptor_t = detail::descriptor_base<task_t>;
+
+template <typename Float>
+std::int64_t train_kernel_cov_impl<Float>::get_row_block_count(std::int64_t row_count) {
+    ONEDAL_ASSERT(row_count > 0);
+    // TODO optimize the approach for row_block_count calculating
+
+    std::int64_t row_block_count = 128;
+    if (row_count < 5000)
+        row_block_count = 1;
+    else if (row_count < 10000)
+        row_block_count = 8;
+    else if (row_count < 20000)
+        row_block_count = 16;
+    else if (row_count < 50000)
+        row_block_count = 32;
+    else if (row_count < 100000)
+        row_block_count = 64;
+
+    return row_block_count;
+}
+
+template <typename Float>
+std::int64_t train_kernel_cov_impl<Float>::get_column_block_count(std::int64_t column_count) {
+    ONEDAL_ASSERT(column_count > 0);
+
+    std::int64_t max_work_group_size = dal::backend::device_max_wg_size(q_);
+    return (column_count + max_work_group_size - 1) / max_work_group_size;
```

Comment:
It looks a little bit confusing. Isn't it equivalent to 
```suggestion
    return column_count / max_wg_size + bool(column_count & max_wg_size);
```